lib/src/main/java/com/diffplug/spotless/Jvm.java

```diff
@@ -31,118 +33,208 @@
 		}
 
-		private Support(String formatterName, @Nullable Comparator<? super V> formatterVersionComparator) {
-			java2formatterVersion = new TreeMap<Integer, V>();
-			this.formatterName = formatterName;
-			if (null == formatterVersionComparator) {
-				this.formatterVersionComparator = new SemanticVersionComparator<V>();
-			} else {
-				this.formatterVersionComparator = formatterVersionComparator;
-			}
-			requiredJavaVersion = 0;
-			usingLatestJavaVersion = true;
+		private Support(String formatterName, Comparator<? super V> formatterVersionComparator) {
+			fmtName = formatterName;
+			fmtVersionComparator = formatterVersionComparator;
+			jvm2fmtVersion = new TreeMap<Integer, V>();
+			fmt2jvmVersion = new TreeMap<V, Integer>(formatterVersionComparator);
 		}
 
 		/**
 		 * Add supported formatter version
-		 * @param minimumJavaVersion Minimum Java version required
+		 * @param minimumJvmVersion Minimum Java version required
 		 * @param maxFormatterVersion Maximum formatter version supported by the Java version
 		 * @return this
 		 */
-		public Support<V> add(int minimumJavaVersion, V maxFormatterVersion) {
+		public Support<V> add(int minimumJvmVersion, V maxFormatterVersion) {
 			Objects.requireNonNull(maxFormatterVersion);
-			java2formatterVersion.put(minimumJavaVersion, maxFormatterVersion);
-			requiredJavaVersion = java2formatterVersion.floorKey(Jvm.version());
-			usingLatestJavaVersion &= minimumJavaVersion < Jvm.version();
+			jvm2fmtVersion.put(minimumJvmVersion, maxFormatterVersion);
+			fmt2jvmVersion.put(maxFormatterVersion, minimumJvmVersion);
 			return this;
 		}
 
-		/** @return Latest formatter version supported by this JVM */
-		public V getLatestFormatterVersion() {
-			return java2formatterVersion.get(requiredJavaVersion);
+		/** @throws AssertionError in case the added support entries are illogical */
+		public void verifyConfiguration() {
+			if (jvm2fmtVersion.size() > fmt2jvmVersion.size()) {
+				throw new AssertionError("The added support information contains duplicated formatter versions.");
+			}
+			if (jvm2fmtVersion.size() < fmt2jvmVersion.size()) {
+				throw new AssertionError("The added support information contains duplicated JVM versions.");
+			}
+			jvm2fmtVersion.forEach((jvmVersion, fmtVersion) -> {
+				Map.Entry<Integer, V> lower = jvm2fmtVersion.lowerEntry(jvmVersion);
+				if ((null != lower) && (fmtVersionComparator.compare(fmtVersion, lower.getValue()) <= 0)) {
+					throw new AssertionError(String.format("%d/%s should be lower than %d/%s", jvmVersion, fmtVersion, lower.getKey(), lower.getValue()));
+				}
+				Map.Entry<Integer, V> higher = jvm2fmtVersion.higherEntry(jvmVersion);
+				if ((null != higher) && (fmtVersionComparator.compare(fmtVersion, higher.getValue()) >= 0)) {
+					throw new AssertionError(String.format("%d/%s should be higher than %d/%s", jvmVersion, fmtVersion, higher.getKey(), higher.getValue()));
+				}
+			});
+		}
```

Comment:
@nedtwigg 
I moved unnecessary checks to an external method. My problem with `Jvm.Support` is the static configuration. It fits best with the current formatter step implementations, but it is not lazy. So I prefer to do sanity checks in the UTs.

Reply:
Is your concern about laziness related to performance? I think navigating these tiny treemaps is going to be undetectable in terms of performance. If I understand your concern correctly (which maybe I don't), the tradeoff is:

- we could verify that the `Jvm.Support` metadata is consistent everytime they are used, which means we are adding microseconds of overhead to every single Spotless user forever
- we could add a unit test for every step that uses `Jvm.Support`
  - savings = microseconds of user time per build, rounds down to zero imo
  - costs = at least tens of minutes of contributor and reviewer time per PR

So not a big deal either way, but my preference is for less burden on the programmer if the only cost is microseconds of startup time.

Reply:
Darn, my comment showed up twice, so I deleted one and they both got deleted :(

I think the sanity checks are fast enough that they cost ~0 to users, but adding a chore to the unit tests costs 10s of minutes per PR, so I lean very lightly towards making the verification automatic rather than adding them to a unit test checklist. I'm fine either way though, the core idea here is good.